pkg/controller/infrastructure/flow_reconciler.go

```diff
@@ -0,0 +1,168 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package infrastructure
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/gardener/gardener/extensions/pkg/controller"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/gardener/gardener-extension-provider-gcp/pkg/apis/gcp"
+	"github.com/gardener/gardener-extension-provider-gcp/pkg/controller/infrastructure/infraflow"
+	gcpinternal "github.com/gardener/gardener-extension-provider-gcp/pkg/gcp"
+	gcpclient "github.com/gardener/gardener-extension-provider-gcp/pkg/gcp/client"
+	"github.com/gardener/gardener-extension-provider-gcp/pkg/internal"
+	infrainternal "github.com/gardener/gardener-extension-provider-gcp/pkg/internal/infrastructure"
+)
+
+// FlowReconciler an implementation of an infrastructure reconciler using native SDKs.
+type FlowReconciler struct {
+	client                     client.Client
+	restConfig                 *rest.Config
+	log                        logr.Logger
+	disableProjectedTokenMount bool
+}
+
+// NewFlowReconciler creates a new flow reconciler.
+func NewFlowReconciler(client client.Client, restConfig *rest.Config, log logr.Logger, projToken bool) (Reconciler, error) {
+	return &FlowReconciler{
+		client:                     client,
+		restConfig:                 restConfig,
+		log:                        log,
+		disableProjectedTokenMount: projToken,
+	}, nil
+}
+
+// Reconcile reconciles the infrastructure and returns the status (state of the world), the state (input for the next loops) and any errors that occurred.
+func (f *FlowReconciler) Reconcile(ctx context.Context, infra *extensionsv1alpha1.Infrastructure, cluster *controller.Cluster) error {
+	infraState, err := f.infrastructureStateFromRaw(ctx, infra)
+	if err != nil {
+		return err
+	}
+
+	serviceAccount, err := gcpinternal.GetServiceAccountFromSecretReference(ctx, f.client, infra.Spec.SecretRef)
+	if err != nil {
+		return err
+	}
+
+	fctx, err := infraflow.NewFlowContext(ctx, infraflow.Opts{
+		Log:            f.log,
+		Infra:          infra,
+		State:          infraState,
+		Cluster:        cluster,
+		ServiceAccount: serviceAccount,
+		Factory:        gcpclient.New(),
+		Client:         f.client,
+		PersistFunc: func(ctx context.Context, state *runtime.RawExtension) error {
+			return patchProviderStatusAndState(ctx, f.client, infra, nil, state)
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create flow context: %v", err)
+	}
+
+	status, state, err := fctx.Reconcile(ctx)
```

Comment:
Can we maybe come up with something more meaningful then status and state.
Some suggestions: privState, currentState, progState, progress...